### ISSUE TITLE
Long press in multigesture

### ISSUE THREAD

**Issue by Biep** [koreader/koreader#5534]
An element missing in multigestures is the long press.
I should like to open links to footnotes and references not by a simple tap, because in articles with lots of them I find myself jumping half the time I merely wanted to turn the page.  And sometimes there is a link overlaying a word I want to look up in the dictionary.
So I'd want to long press a word, and then swipe one way to jump, and another to look up the word in the dictionary.  Maybe yet another direction for Wikipedia, and so on.
Especially for starting multigestures the long press would be intuitive (and easy, as one can deliberately position one's finger), but I can imagine uses for ending with a long press, or even having them between swipes.

**Comment by Frenzie** [comment]
You can disable tap on links and only use a multiswipe to follow links. (By default southeast, northeast.)

**Comment by Biep** [comment]
I know.  I was (and am) just hoping for another enhancement to multigestures.

**Comment by Frenzie** [comment]
Long-press on corner is now available in gesture manager.

![image](https://user-images.githubusercontent.com/202757/220997011-54f55488-d535-42ee-ae71-f7ee292369f5.png)

**Comment by Biep** [comment]
Thanks!
But I was more thinking of having long presses as part of multigestures.  Say, a long press followed by an upswipe to move scroll the point pressed to the top of the screen, or by a downswipe to scroll it to the bottom of the screen.  (When reading mathematical proofs it often helps enormously to have the whole proof visible at once, instead of repeatedly paging fore and back.)

**Comment by Frenzie** [comment]
Doesn't that describe how panning already works? :/

**Comment by poire-z** [comment]
Hold, pan, pan_release, hold_release... anything with long-press at start or end of the gesture are handled specifically by the different widgets being the top widget (in Reader, for scrolling/panning, in other widgets by MovableContainer), so it would be really hard to be able to abstract them in something as generic as our Gesture manager.
Moreover, the different actions you described and would like gestures for are not really generic (we don't have such specific actions depending on the location of the start or end or the distance - except may be swipe on the edges). They would have to be handled by the different modules (readerlink, readerscroll) - and there's probably not many other gesture than the one you described that would be worth associating to these new actions :) (Long press on a link and pan left won't be used to open History :)
So, your suggestions would need to be implemented as options to the already existing features and handlers (ie. in ReaderHighlight, if hold + pan up, launch Wikipedia, and have various direct actions associated to each of the 4 N/W/S/E directions). But then I see already a conflict: hold + pan up selects multiple lines, so your idea is killing non-single-word selection.)

TLDR: probably too complicated (and impacting the existing code) to be implemented generically.
Probably easier to implement invidividual small ideas around what exists, if they don't cause conflicts.

**Comment by Biep** [comment]
> But then I see already a conflict: hold + pan up selects multiple lines, so your idea is killing non-single-word selection.

But precisely *that* would be up to the gesture settings, wouldn't it?  Currently Hold+pan to X selects from the hold position to the pan end position - but that would be overridable.  Maybe I would prefer the inverse there: pan+hold, for instance, or hold-pan-hold (my actual preference).

The start could be merely to refactor these gestures out of the modules and into the manager, each gesture moved that way should simplify the code *and* add to genericity.  How many such hardcoded gestures are there?

**Comment by poire-z** [comment]
Not really. For these gestures we are talking about, there is a state and things happening while it is being made (redrawing the page, moving a widget).
It's simpler to have all that handled in the dedicated module, than having abstract half-gestures broadcasted to all the code - and some bits of them overridden by crazy users :) and breaking all the states.
Really not caring about how many, or shuffling all that around while it's not broken (and took years to get it to a non-broken state :)

**Comment by jonnyl2** [comment]
@poire-z or @Frenzie: not related to long press, but multiswipes: have two-finger multiswipes ever been considered? The reason I'm asking: although there's almost limitless options for gestures with 3 or more legs, the time to draw them and the effort to remember them is the true limitation. Two-finger multiswipe gestures would allow more options with shorter execution time and they would be easier to remember. E.g., two fingers up/down, left/right, northwest-southeast etc. I've wanted to ask about this for a long time and thought this was a good opportunity :)

**Comment by poire-z** [comment]
My experience with 2-fingers swipes is bad. At the time, before multiswipes and gestures manager were implemented, it was the option to open ToC and Bookmarks (via 2-fingers swipe east/west - might still work :). And half of the times I did these gesture, I didn't trigger anything, or it was recognized as a single swipe and changed page.
So, personally not caring any longer about two-finger swipes :)

**Comment by NiLuJe** [comment]
They can be iffy on touchy (pun intended) panels, although they're hopefully a bit more reliable since the GestureManager rework.

**Comment by Frenzie** [comment]
I think they work alright for simple gestures but for more complex it's a pain to maneuver your fingers right.

**Comment by jonnyl2** [comment]
Actually, now I remember that in the beginning I used to experience some frustration with two-finger swipes too, but then I adapted by spreading the two fingers a bit wider and that pretty much resolved it. I think for just two-legged multiswipes they would be nice to have, and should be easy to maneuver, but if there's no real demand for it then that's okay too. Can't really complain about KOReader's gesture features :)
